### PR TITLE
doc: New performance benchmarks and tuning guide

### PR DESCRIPTION
**Pull request by tgraf**
Updated performance benchmark section and a brand new tuning guide
explaining how to achieve these results.

### CODE REVIEW

File: Documentation/operations/performance/benchmark.rst [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+.. _performance_report:
+
+**********************
+Performance Benchmarks
+**********************
+
+Introduction
+============
+
+The following section contain performance benchmark numbers of various
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst]:
```suggestion
The following section contains performance benchmark numbers of various
```

File: Documentation/operations/performance/benchmark.rst [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+Single-Stream
+-------------
+
+In this test, a single TCP stream is opened between the pods and maximum
+throughput is achieved:
+
+.. image:: images/bench_tcp_stream_1_stream.png
+
+When running in the fastest datapath option,
+Cilium can outperform even the node to node baseline even though it is
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst]:
```suggestion
Cilium can outperform even the node-to-node baseline even though it is
```

File: Documentation/operations/performance/benchmark.rst [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+performing more work (forwarding into the network namespace of the Kubernetes
+pod). This is because Cilium is capable of bypassing the iptables layer of the
+node which is still traversed for node to node baseline.
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst]:
```suggestion
node which is still traversed for node-to-node baseline.
```

File: Documentation/operations/performance/benchmark.rst [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+
+.. image:: images/bench_tcp_stream_32_streams.png
+
+Given multiple processes are being used, all test configurations can achieve
+close to the line-rate of the NIC. The main difference is the CPU resources
+required to achieve it:
+
+.. image:: images/bench_tcp_stream_32_streams_cpu.png
+
+Considerable more CPU is spent in total. This is because more processes are
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst]: > Considerable more CPU is spent in total.

I would make explicit more than _what_.

File: Documentation/operations/performance/benchmark.rst [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+running and more context switching is required.
+
+Request/Reply Rate
+==================
+
+The request/reply rate primarily measures the latency and efficient to handle
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst]:
```suggestion
The request/reply rate primarily measures the latency and efficiency to handle
```

File: Documentation/operations/performance/benchmark.rst [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+
+This test represents a workload that receives or initiates a lot of unique TCP
+connections. This could be the case for a publicly exposed service that
+receives a lot of client connections such as a L4 proxy or a service opening
+connections to many external endpoints. This benchmark puts the most stress on
+the actual systems compared with the fewest work being offloaded to hardware so
+we can expect to see the biggest differences in the benchmark results.
+
+A configuration that does well in this test (delivering high connection rates)
+will handling situation with overwhelming connection rates better, leaving more
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst]:
```suggestion
will handle situation with overwhelming connection rates better, leaving more
```

File: Documentation/operations/performance/benchmark.rst [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+32 Processes
+------------
+
+In this test, 32 processes running in parallel open as many TCP connections in
+sequence as possible. This is by far the most stressful test for the system.
+
+.. image:: images/bench_tcp_crr_32_processes.png
+
+This benchmark outlines major differences between the tested configurations. It
+in particular illustrates the overall cost of iptables which is optimized to
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst:196]:
```suggestion
This benchmark outlines major differences between the tested configurations. In
particular, it illustrates the overall cost of iptables which is optimized to
```

File: Documentation/operations/performance/benchmark.rst [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+**********************
+Performance Benchmarks
+**********************
+
+Introduction
+============
+
+This chapter contains performance benchmark numbers for various scenarios. All
+tests are performed between containers running on two different bare metal
+nodes connected by a 100Gbit network interface.
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst:18]:
```suggestion
nodes connected by a 100Gbit/s network interface.
```

File: Documentation/operations/performance/benchmark.rst [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+This test represents a workload that receives or initiates a lot of TCP
+connections. An example where this is the case is a publicly exposed service
+that receives connections from many clients. Good examples of this are L4
+proxies or services opening many connections to external endpoints. This
+benchmark puts the most stress on the system with the least work offloaded to
+hardware so we can expect to see the biggest difference between tested
+configurations.
+
+A configuration that does well in this test (delivering high connection rates)
+will handling situation with overwhelming connection rates much better, leaving
```
Comment (qmonnet) [inline Documentation/operations/performance/benchmark.rst:164]:
```suggestion
will handle situation with overwhelming connection rates much better, leaving
```

File: Documentation/operations/performance/tuning.rst [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+             --namespace kube-system \\
+             --set hubble.enabled=false
+
+MTU
+===
+
+The maximum transfer unit (MTU) can have a significant impact on the network
+throughput of a configuration. Cilium will automatically detect the MTU of the
+underlying network devices. Therefore, if your system is configured to use
+jumbo frames then Cilium will automatically make us of it.
```
Comment (qmonnet) [inline Documentation/operations/performance/tuning.rst:100]:
```suggestion
jumbo frames then Cilium will automatically make use of it. 
```

File: Documentation/operations/performance/tuning.rst [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+jumbo frames then Cilium will automatically make us of it. 
+
+To benefit from this, make sure that your system is configured to use jumbo
+frames if your network allows for it.
+
+Kernel Optimizations
+====================
+
+The kernel allows to configure several options which will help maximize network
+performance:
```
Comment (qmonnet) [inline Documentation/operations/performance/tuning.rst:109]: Would be interesting for users, I think, to have a one-line description of what each optimisation does.

**Review by qmonnet — CHANGES_REQUESTED** [review]
A few typos & minor suggestions.

File: Documentation/operations/performance/benchmark.rst (hunk 1 above)
Comment (qmonnet) [reply Documentation/operations/performance/benchmark.rst]: Sorry about that - I thought this comment had been automatically removed when I refreshed the page on the latest commit :/.

File: Documentation/operations/performance/benchmark.rst [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+  a single request/reply payload message transmitted for each new connection. A
+  single process and 32 parallel processes are tested.
+
+.. _benchmark_throughput:
+
+Throughput
+==========
+
+Throughput testing is useful to understand the maximum throughput that can be
+achieved with a particular configuration. All or most configuration will be
```
Comment (lizrice) [inline Documentation/operations/performance/benchmark.rst:52]:
```suggestion
achieved with a particular configuration. All or most configurations will be
```

File: Documentation/operations/performance/benchmark.rst [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+Cilium (legacy host-routing) Cilium 1.9.6, legacy host-routing, kube-proxy replacement, No CT
+Calico                       Calico 3.17.3, kube-proxy
+Calico eBPF                  Calico 3.17.3, eBPF datapath, No CT
+============================ ===================================================================
+
+How to reproduce
+================
+
+To ease reproducibility, this report is paired with a set of scripts that can be
+found in https://github.com/cilium/perfeval. All scripts in this document refer
```
Comment (aanm) [inline Documentation/operations/performance/benchmark.rst:252]: can be found in https://github.com/cilium/cilium-perf-networking

**Review by aanm — APPROVED** [review]
Just one small nit. LGTM

File: Documentation/operations/performance/benchmark.rst [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+
+Cilium achieves almost the same request/reply rate as the baseline while only
+consuming marginally more CPU resources:
+
+.. image:: images/bench_tcp_rr_1_process_cpu.png
+
+32 Processes
+------------
+
+In this test, 32 processes are opening 32 parallel TCP connection. Each process
```
Comment (ianvernon) [inline Documentation/operations/performance/benchmark.rst:133]: connection --> connections

**Review by ianvernon — COMMENTED** [review]
Nice work! Just one small grammatical nit

**Comment by borkmann** [comment]
@ianvernon @aanm @lizrice Note that this PR still needs to switch its base to https://github.com/cilium/cilium/commits/pr/tgraf/performance-docs given we've pushed various improvements on top which should at least address some of your feedback.

**Comment by tgraf** [comment]
Superseded by #15943

**Pull request closed without merging**